validator/core/src/lib.rs

```diff
@@ -0,0 +1,117 @@
+//! Core library for validating [Substrait](https://substrait.io/).
+//!
+//! Note that this crate is simply the greatest common denominator of the
+//! various language bindings for the validator, which doesn't include the
+//! remote YAML file resolution due to the complexities of linking against
+//! the system's cryptography/certificate validation libraries. Refer to
+//! the documentation for the `substrait_validator` crate for basic usage
+//! from within Rust.
+
+#[macro_use]
+pub mod output;
+
+#[macro_use]
+mod parse;
+
+pub mod export;
+pub mod input;
+
+// Aliases for common types used on the crate interface.
+pub type Config = input::config::Config;
+pub type Pattern = input::config::glob::Pattern;
+pub type Diagnostic = output::diagnostic::Diagnostic;
+pub type Classification = output::diagnostic::Classification;
+pub type Level = output::diagnostic::Level;
+pub type Node = output::tree::Node;
+
+/// Validates the given substrait.Plan message and returns the parse tree.
+pub fn parse<B: prost::bytes::Buf>(buffer: B, config: &Config) -> Node {
+    parse::parse(buffer, config)
+}
+
+/// Iterates over all diagnostic messages in the tree.
+pub fn iter_diagnostics(root: &Node) -> impl Iterator<Item = &Diagnostic> + '_ {
```

Comment:
Why not make this and the other methods members of Node?

Reply:
I'm kind of on the fence about this. I like the idea of putting all the functions you would need as a normal user to get basic validation going in the toplevel module of the crate, so it's easy to find where the entry points for the library are. But maybe, like `iter_diagnostic()`, they should all *also* be methods of `Node`.

I could also make the root of the tree an explicit type at the top level of the crate and implement these methods on it, since these methods are only really intended to be used for a tree as a whole, rather than on any subtree. But that doesn't feel right either, since it's not necessarily wrong to apply them to subtrees, either.

Maybe both? It hasn't been feeling right to pass `Node` around as the complete representation of a tree. In the C and Python bindings I needed wrappers anyway, where I called the wrapper `substrait_validator_result_handle` and `ParseResult` respectively. Might as well do the same in Rust, even if it would only be a thin wrapper there (unless at some point additional root-only info needs to be passed, so it might be a good idea to make a struct for that already for forward-compatibility's sake, especially in the proto serialization of this).

Guess I have to mess around with this a bit more.